### PR TITLE
return 503 for aggregated APIs when the APIServiceRegistrationController hasn't finished installing all known APIServices

### LINKED ISSUE #104342: GC might delete objects during kube-apiserver startup
#### What happened:
GC may delete objects during startup while the apiserver has not fully initialized yet.

One potential case where it can happen: CRDs are available through informers 
- kube-apiserver is starting, its informers have not synced yet
- GC sends a query (GET a CR)
- since the informers have not synced yet, GC may get a `404`
- This may potentially lead GC to delete objects 

#### What you expected to happen:
`kube-apiserver` should respond with a `Retry-After` until it has fully initialized. GC will get a `429` with a `Retry-After` response header in this case.

#### How to reproduce it (as minimally and precisely as possible):

With an HA cluster it's hard to reproduce.

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it: 
Ensures that requests to aggregated APIs will get 503, not 404 while the APIServiceRegistrationController hasn't finished installing all known APIServices. It is especially important for controllers like GC and NS since they act on 404

Note that this is an already existing behavior (returning 503 when the CRD informer hasn't been sync). 
We already have it for the apiextentionserver, see https://github.com/kubernetes/kubernetes/pull/105653 for more details.

#### Which issue(s) this PR fixes:

Fixes #104342

#### Does this PR introduce a user-facing change?

```release-note
resolves a potential issue with GC and NS controllers which may delete objects after getting a 404 response from the server during its startup. This PR ensures that requests to aggregated APIs will get 503, not 404 while the APIServiceRegistrationController hasn't finished its job.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ package app
 import (
 	"crypto/tls"
 	"fmt"
+	"k8s.io/apiserver/pkg/util/notfoundhandler"
```
Comment (sttts) [inline cmd/kube-apiserver/app/server.go:25]: order

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 2]
```diff
@@ -271,14 +271,19 @@ func (s *GenericAPIServer) NextDelegate() DelegationTarget {
...
 	return emptyDelegate{}
 }
 
+func NewEmptyDelegateWithCustomHandler(handler http.Handler) DelegationTarget {
+	return emptyDelegate{handler}
+}
+
 func (s emptyDelegate) UnprotectedHandler() http.Handler {
-	return nil
+	return s.handler
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:286]: remind me why we only need this for the unprotected handler

File: staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/util/signal"
+)
+
+// New returns an HTTP handler that is meant to be executed at the end of the delegation chain and check if the server has been fully prepared before replying to the client.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go:32]: mention that this is long before readiness and is meant to block request before the muxes and handlers are completely initialized such that requests are properly routed.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go:32]: And that preparedToServe must not depend on other components in the cluster other than etcd

File: staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+type Handler struct {
+	preparedToServeSignals []signal.Receiver
+	serializer             runtime.NegotiatedSerializer
+}
+
+func (h *Handler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	for _, preparedToServeSignal := range h.preparedToServeSignals {
+		if !preparedToServeSignal.IsSignaled() {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/util/notfoundhandler/not_found_handler.go:45]: continue early

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:286]: this is the one we use at the end of the chain, here https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L190

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 5]
```diff
@@ -113,10 +114,26 @@ type preparedAPIAggregator struct {
 	runnable runnable
 }
 
+// preparedToServeSignals holds signals/events that indicate this server has been fully initialized and is ready to serve requests
+type preparedToServeSignals struct {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:118]: why a wrapper around another type? Just use `signal.Interface` directly

Comment (p0lyn0mial) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:118]: I wanted to expose only the read-only interface

**Comment by p0lyn0mial** [comment]
I will remove the signal pkg and use raw types instead. I will also add a signal from the extension API server.

File: staging/src/k8s.io/apiserver/pkg/util/signal/signal.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	name string
+	ch   chan struct{}
+}
+
+func (e *namedChannelWrapper) Signal() {
+	select {
+	case <-e.ch:
+		// already closed, don't close again.
+	default:
+		close(e.ch)
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/util/signal/signal.go:44]: if `Signal` is used  concurrently, i think we need to use a `sync.Once` to protect it.

File: staging/src/k8s.io/apiserver/pkg/util/signal/signal.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func New(name string) Interface {
+	return &namedChannelWrapper{
+		name: name,
+		ch:   make(chan struct{}),
+	}
+}
+
+type namedChannelWrapper struct {
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/util/signal/signal.go:34]: since we are converting this to a shared utility, i would recommend unit tests for it.

**Comment by caesarxuchao** [comment]
/cc @cheftako @jpbetz 

Walter, I think you mentioned a specific http status code during the sig meeting.

**Comment by tkashem** [comment]
cc @MikeSpreitzer

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 2 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:286]: Can you please help me understand, by spelling out the rest of the answer?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 8]
```diff
@@ -271,14 +271,19 @@ func (s *GenericAPIServer) NextDelegate() DelegationTarget {
 }
 
 type emptyDelegate struct {
+	handler http.Handler
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:274]: What did "empty delegate" mean, exactly?  Is it still empty?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:118]: But the current text exposes the full interface.
It looks to me like the reason for a struct here is to prepare for more signals in the future.  For example, one about RBAC being initialized.

**Comment by MikeSpreitzer** [comment]
We discussed this in the Aug 25 SIG api-machinery meeting.  If I recall correctly, the consensus was that the client should be made to opt in to this.  We have started drafting a KEP, at https://docs.google.com/document/d/1JQNT0eu6Gpn9UZ4rcfIUDkJf1wKc3f_57KKC4c-_s-U .

**Comment by MikeSpreitzer** [comment]
Also, we want this conditionality to cover RBAC initialization too.  A too-early answer in that regard is an authorization failure rather than a 404.

**Comment by cheftako** [comment]
> /cc @cheftako @jpbetz
> 
> Walter, I think you mentioned a specific http status code during the sig meeting.

Partly depends on the idea. 
I liked the precondition solution in which case I think 412 would be the correct response code.
It seemed a more backward compatible fix in that clients would clearly be opting in.

However this seems to be a general "KAS not ready yet" solution.
For that I think 503 is the correct response node.

**Comment by sttts** [comment]
Lgtm and approved.

Will give @MikeSpreitzer @caesarxuchao and others a chance to take a final look.

**Review by MikeSpreitzer — COMMENTED** [review]
What about https://docs.google.com/document/d/1JQNT0eu6Gpn9UZ4rcfIUDkJf1wKc3f_57KKC4c-_s-U ?  My recollection of the SIG api-machinery discussion is that a client opt-in was favored, out of concern that a non-optional solution would ruining some startup sequences that involve multiple processes.

**Comment by p0lyn0mial** [comment]
@sttts @MikeSpreitzer many thanks for reviewing the PR. I applied the last batch of comments. I think we are ready to tag.

**Review by MikeSpreitzer — COMMENTED** [review]
This looks basically OK to me.
I have just a few minor comments.

**Pull request merged**